ambry-cloud/src/main/java/com/github/ambry/cloud/CloudBlobStore.java

```diff
@@ -416,16 +416,21 @@ public short undelete(MessageInfo info) throws StoreException {
    * @param lifeVersion life version of the deleted blob.
    * @return final updated life version of the blob.
    * @throws CloudStorageException in case any exception happens during undelete.
+   * @throws StoreException in case any {@link StoreException} is thrown.
    */
-  private short undeleteIfNeeded(BlobId blobId, short lifeVersion) throws CloudStorageException {
+  private short undeleteIfNeeded(BlobId blobId, short lifeVersion) throws CloudStorageException, StoreException {
     // See note in deleteIfNeeded.
     if (!checkCacheState(blobId.getID(), lifeVersion, BlobState.CREATED)) {
       short newLifeVersion = cloudDestination.undeleteBlob(blobId, lifeVersion, this::preUndeleteValidation);
       addToCache(blobId.getID(), newLifeVersion, BlobState.CREATED);
       return newLifeVersion;
     } else {
-      throw new CloudStorageException("Error updating blob metadata",
-          new StoreException("Id " + blobId.getID() + " is already undeleted in cloud", StoreErrorCodes.ID_Undeleted));
+      if (lifeVersion > 0) {
+        throw new StoreException("Id " + blobId.getID() + " is already undeleted in cloud",
+            StoreErrorCodes.ID_Undeleted);
+      } else {
+        throw new StoreException("Id " + blobId.getID() + " not deleted yet in cloud", StoreErrorCodes.ID_Not_Deleted);
```

Comment:
Is it possible that is has been deleted in the cloud but didn't go through this vcr so the cache entry is stale?